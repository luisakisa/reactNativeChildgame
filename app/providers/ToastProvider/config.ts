export const TOAST_LIFETIME_DURATION = 3000;
