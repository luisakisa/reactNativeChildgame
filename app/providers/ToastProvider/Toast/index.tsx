import React, {memo, useContext, useEffect} from 'react';
import Animated, {
  SharedValue,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withDelay,
  withTiming,
  cancelAnimation,
  withSequence,
} from 'react-native-reanimated';
import styles from './styles';
import {LayoutChangeEvent, Text} from 'react-native';
import {TOAST_LIFETIME_DURATION} from '../config';
import {AppThemeContext} from '../../AppTheme';

const PRESENTATION_DURATION = 300;
const HIDE_DURATION = 700;

interface Props {
  message: string;
  index: number;
  toastsLength: SharedValue<number>;
}

function Toast({message, index, toastsLength}: Props): React.FunctionComponentElement<Props> {
  const theme = useContext(AppThemeContext);

  const scale = useSharedValue<number>(0.7);
  const opacity = useSharedValue<number>(0);
  const height = useSharedValue<number>(0);
  const activeToast = useSharedValue<boolean>(true);

  useEffect(() => {
    scale.value = withTiming(1, {duration: PRESENTATION_DURATION});
    opacity.value = withSequence(
      withTiming(1, {duration: PRESENTATION_DURATION}),
      withDelay(
        TOAST_LIFETIME_DURATION - PRESENTATION_DURATION - HIDE_DURATION - 100,
        withTiming(0, {duration: HIDE_DURATION}),
      ),
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const containerAnimatedStyle = useAnimatedStyle(() => {
    return {
      height: withTiming(height.value + (height.value > 0 ? 8 : 0), {
        duration: PRESENTATION_DURATION,
      }),
    };
  }, [height]);

  const toastAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: opacity.value,
      transform: [{scale: scale.value}],
    };
  }, [scale, opacity]);

  function handleLayout(event: LayoutChangeEvent): void {
    height.value = event.nativeEvent.layout.height;
  }

  useDerivedValue(() => {
    if (toastsLength.value !== 0 && index === toastsLength.value - 2 && activeToast.value) {
      activeToast.value = false;
      cancelAnimation(opacity);
      opacity.value = withDelay(
        PRESENTATION_DURATION + 200,
        withTiming(0, {duration: HIDE_DURATION}),
      );
    }
  }, [toastsLength]);

  return (
    <Animated.View style={[styles.container, {zIndex: 999 - index}, containerAnimatedStyle]}>
      <Animated.View
        style={[styles.toast, {backgroundColor: theme.C15}, toastAnimatedStyle]}
        onLayout={handleLayout}
      >
        <Text style={[styles.message, {color: theme.C12}]}>{message}</Text>
      </Animated.View>
    </Animated.View>
  );
}

export default memo(Toast);
