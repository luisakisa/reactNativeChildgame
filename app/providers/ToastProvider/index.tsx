import React, {createContext, useEffect, useState} from 'react';
import {useSharedValue} from 'react-native-reanimated';
import styles from './styles';
import {View} from 'react-native';
import {useSafeAreaInsets} from 'react-native-safe-area-context';
import Toast from './Toast';
import {TOAST_LIFETIME_DURATION} from './config';
import {Mutex} from 'async-mutex';
import {useDebounce} from '../../hooks/useDebounce';

const mutex = new Mutex();

interface ToastProviderType {
  toast: (message: string) => void;
}

export const ToastContext = createContext<ToastProviderType>({
  toast: (): void => {},
});

export default function ToastProvider(props: any): JSX.Element {
  const safeAreaInsets = useSafeAreaInsets();

  const [toasts, setToasts] = useState<string[]>([]);

  const toastsLength = useSharedValue<number>(0);

  useEffect(() => {
    toastsLength.value = toasts.length;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [toasts]);

  function showToast(message: string): void {
    mutex.acquire().then(release => {
      setToasts(prev => [...prev, message]);
      setTimeout(release, 500);
    });
  }

  useDebounce(
    () => {
      if (toasts.length > 0) {
        setToasts([]);
      }
    },
    TOAST_LIFETIME_DURATION,
    [toasts],
  );

  return (
    <ToastContext.Provider value={{toast: showToast}}>
      {props.children}
      <View style={[styles.container, {bottom: safeAreaInsets.bottom + 24}]}>
        {toasts.map((message, index) => {
          return (
            <Toast
              key={`toast-${index}`}
              message={message}
              index={index}
              toastsLength={toastsLength}
            />
          );
        })}
      </View>
    </ToastContext.Provider>
  );
}
