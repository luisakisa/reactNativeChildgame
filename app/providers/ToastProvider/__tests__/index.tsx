import React from 'react';
import {render} from '@testing-library/react-native';
import ToastProvider from '../index';

test('provider/ToastProvider/snapshot', () => {
  const {toJSON} = render(<ToastProvider />);
  expect(toJSON()).toMatchSnapshot();
});
