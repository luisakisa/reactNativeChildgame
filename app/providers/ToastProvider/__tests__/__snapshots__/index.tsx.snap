// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider/ToastProvider/snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
      },
      Object {
        "bottom": 24,
      },
    ]
  }
/>
`;
